Fix #785, Fix #1441, prevent race condition on `interrupted` variable by moving `interrupted` into atomic state

neko-kai opened the pull request:
Also, prevent potential volatility problems with accessing interruptStack from multiple threads by moving `interruptible` state into Suspended

neko-kai commented:
A clean reproduction for testing:

```scala
import java.util.concurrent.atomic.{AtomicLong, LongAdder}

import zio.{UIO, _}

import scala.language.reflectiveCalls

object ZioInterruptLeakOrDeadlockRepo extends zio.App {

  val leakedCounter = new AtomicLong(0L)

  val startedCounter = new LongAdder
  val completedCounter = new LongAdder
  val awakeCounter = new LongAdder
  val pendingGauge = new LongAdder

  def run(args: List[String]): ZIO[Environment, Nothing, Int] = {
    val leakOrDeadlockTest = for {
      _ <- UIO {
        startedCounter.increment()
        pendingGauge.increment()
      }

      sleepInterruptFiber <- IO.never.fork

      // This blocks / leaks once every 20,000 rounds or so
      _ <- sleepInterruptFiber.interrupt
        .ensuring(UIO {
          awakeCounter.increment()
          pendingGauge.decrement()
        })

      _ <- UIO(completedCounter.increment())
    } yield ()

    val main = for {
      _ <-
        ZIO.runtime[Any].map(_.Platform.executor.metrics.get)
          .flatMap {
            metrics => UIO(new Thread(() => {
              while (true) {
                println(s"started=${startedCounter.longValue()} awake=${awakeCounter.longValue()} completed=${completedCounter.longValue()} pending=${pendingGauge.longValue()} queued=${metrics.size}")
                Thread.sleep(1000L)
              }
            }).start())
          }

      _ <- leakOrDeadlockTest.forever
    } yield 0

    main
  }
}

```

regiskuckaertz commented:
wow, you found it! Awesome detective work 🕵

neko-kai commented:
Race scenario:
```
fiber1: kill0 sets from Running to Running, 
fiber2: BEFORE interruptible=false, but AFTER atomic update, sets state=Suspended and exits.
fiber1: Exits because it thinks fiber2 is Running from previous state
```
@regiskuckaertz 
Seems John also found it at the same time, you could say it was a race :D Let's wait for John's feedback first!

jdegoes commented on core/shared/src/main/scala/zio/internal/FiberContext.scala:
```diff
@@ -675,9 +676,8 @@ private[zio] final class FiberContext[E, A](
-      case Executing(FiberStatus.Suspended, observers) =>
-        if (!state.compareAndSet(oldState, Executing(FiberStatus.Running, observers)))
-          exitAsync()
+      case Executing(_: FiberStatus.Suspended, observers, interrupt) =>
```
How about `FiberStatus.Suspended(_)`, to avoid the appearance of type casing?

neko-kai replied:
I'm not sure if that would result in a call to .unapply, avoiding that is the only reason I used a type match here, will check...

jdegoes replied:
Ah, performance is always a reason to break rules in ZIO. 😆

neko-kai replied:
Hmm, I'm wrong! Scala is smart enough to avoid `unapply` not only here, but everywhere in `FiberContext` there are no unapply calls:
```java
if (var6 instanceof FiberContext$FiberStatus$Suspended) {
   if (!this.state.compareAndSet(oldState, new FiberContext$FiberState$Executing(FiberContext$FiberStatus$Running$.MODULE$, observers, interrupt))) {
      continue;
   }

   var2 = true;
   break;
}
```

jdegoes commented on core/shared/src/main/scala/zio/internal/FiberContext.scala:
```diff
@@ -743,28 +742,23 @@ private[zio] final class FiberContext[E, A](
         }
 
-      case Executing(status, observers0) =>
+      case Executing(status, observers0, _) =>
```
This extra flag is necessary in your fix. In my fix the model changed a bit, so pushing the "set" prior to the atomic ref update should in theory take care of it. But this looks good for now!

jdegoes commented:
@neko-kai Can you add the test?

neko-kai commented:
@jdegoes I've added a test that just runs IO.fork.interrupt 10K times, multiplied by test builds it should probably be able to catch a regression over many builds.

I've also added a benchmark that does the same IO.fork.interrupt and should definitely start hanging on regression. I've also added a benchmark for `race` that highlights the perf issue in #1441 - our `race` is quite a bit slower 😱 
```
[info] Benchmark                                (size)   Mode  Cnt    Score     Error  Units
[info] IOEmptyRaceBenchmark.catsEmptyRace         1000  thrpt   10  315.196 ± 398.678  ops/s
[info] IOEmptyRaceBenchmark.monixEmptyRace        1000  thrpt   10  192.894 ±   7.672  ops/s
[info] IOEmptyRaceBenchmark.zioEmptyRace          1000  thrpt   10   33.152 ±   0.572  ops/s
[info] IOEmptyRaceBenchmark.zioTracedEmptyRace    1000  thrpt   10   25.166 ±   1.548  ops/s
```

neko-kai commented:
I believe another inconsistency with `exitAsync()` can happen as witnessed by the following code:
```scala
IO.effectAsyncMaybe[Nothing, Unit] {
  k =>
    Future {
      Thread.sleep(200L)
      k(IO.dieMessage("INCOHERENT CONTINUE"))
    }
    Some(IO.unit)
}
  .flatMap(_ => ZIO.never)
  .ensuring(IO.dieMessage("CAN'T HAPPEN!"))
  .uninterruptible
```

that "overtakes" the fiber from an async that was "cancelled", but only if the fiber enters suspension at an opportune time

neko-kai commented:
There's also a remaining race with `evaluateLater(ZIO.interrupt)` in `kill0`, illustration:
```scala
object ZioIncoherentInterruptAsyncRepro extends zio.App {

  def run(args: List[String]): ZIO[Environment, Nothing, Int] = {
    val test =
      ZIO.effectAsyncMaybe[Environment, Nothing, Unit] {
        k =>
          Future {
            Thread.sleep(200L)
            k(console.putStrLn("INCOHERENT CONTINUE"))
          }
          Some(IO.unit)
      }
        .flatMap(_ => ZIO.sleep(500.millis))
        .ensuring(console.putStrLn("Async finalizer!") *> ZIO.never)
        .ensuring(console.putStrLn("CAN'T HAPPEN!"))

    test.fork.flatMap {
      f =>
        f.interrupt.delay(50.millis)
    }.as(0)
  }
}
```

Basically, every single point where multi-threaded access can happen in non-atomic or non-guarded way is a problem (:
Luckily, we only really have async + interrupt as these points.

neko-kai commented:
Will address these two - https://github.com/zio/zio/pull/1792#issuecomment-535262826, https://github.com/zio/zio/pull/1792#issuecomment-535266947 - separately.

iravid commented:
@neko-kai Is the performance of race improved after these fixes?

neko-kai commented:
@iravid 
Nope, I just added a benchmark that showcases the bad perf. Will open an issue about race perf too!

iravid commented:
Ah too bad. Ok :-)

neko-kai commented:
@iravid @jdegoes 
Race perf moved here – https://github.com/zio/zio/issues/1844
More async races - https://github.com/zio/zio/issues/1842, https://github.com/zio/zio/issues/1843
